packages/salesforcedx-vscode-apex/src/views/testOutline.ts

```diff
@@ -0,0 +1,666 @@
+      } else {
+        const emptyArray = new Array<ApexTest>();
+        emptyArray.push(new ApexTest('No tests in folder', null, 0));
+        return Promise.resolve(emptyArray);
+      }
+    }
+  }
+
+  public getTreeItem(element: Test): vscode.TreeItem {
+    if (element) {
+      return element;
+    } else {
+      this.getAllApexTests(this.path);
+      if (!(this.head && this.head.children.length > 0)) {
+        this.head = new ApexTest('No tests in folder', null, 0);
+        this.head.children.push(new ApexTest('No tests in folder', null, 0));
+      }
+      return this.head;
+    }
+  }
+
+  public async refresh() {
+    this.head = null; // Reset tests
+    this.apexTestMap.clear();
+    ApexTestOutlineProvider.testStrings.clear();
+    this.apexTestInfo = null;
+    if (isLanguageClientReady()) {
+      this.apexTestInfo = (await getApexTests()) as ApexTestRequestInfo[];
+    }
+    this.apexClasses = await vscode.workspace.findFiles('**/*.cls');
+    this.getAllApexTests(this.path);
+    this._onDidChangeTreeData.fire();
+  }
+
+  private getAllApexTests(path: string): Test {
+    if (this.head == null) {
+      // Starting Out
+      this.head = new ApexTestGroup('ApexTests', null, 1);
+    }
+    // Check if lsp worked, else parse files manually
+    if (this.apexTestInfo) {
+      // Do it the easy way
+      this.apexTestInfo.forEach(test => {
+        let apexGroup = this.apexTestMap.get(test.parent) as ApexTestGroup;
+        if (!apexGroup) {
+          apexGroup = new ApexTestGroup(test.parent, null, 1);
+          this.apexTestMap.set(test.parent, apexGroup);
+        }
+        const testUri = vscode.Uri.file(test.file.substring(7));
+        const apexTest = new ApexTest(test.methodName, testUri, test.line);
+        apexTest.name = apexGroup.label + '.' + apexTest.label;
+        this.apexTestMap.set(apexTest.name, apexTest);
+        apexGroup.children.push(apexTest);
+        if (this.head && !(this.head.children.indexOf(apexGroup) >= 0)) {
+          this.head.children.push(apexGroup);
+        }
+        ApexTestOutlineProvider.testStrings.add(apexGroup.name);
+      });
+    } else {
+      this.apexClasses.forEach(apexClass => {
+        const fileContent = fs.readFileSync(apexClass.fsPath).toString();
+        if (fileContent && fileContent.toLowerCase().includes('@istest')) {
+          const testName = ospath
+            .basename(apexClass.toString())
+            .replace('.cls', '');
+          const newApexTestGroup = new ApexTestGroup(testName, apexClass, 1);
+          this.apexTestMap.set(testName, newApexTestGroup);
+          this.addTests(fileContent, newApexTestGroup);
+          ApexTestOutlineProvider.testStrings.add(testName);
+          if (this.head) {
+            this.head.children.push(newApexTestGroup);
+          }
+        }
+      });
+    }
+    return this.head;
+  }
+
+  private addTests(fileContent: string, apexTestGroup: ApexTestGroup): void {
```

Comment:
Do we really need this? It seems like this is very error-prone since it's not actually parsing but matching regex. You could be confused if you have a commented portion of code like

```
// @IsTest
```

or

```
/**
* @IsTest
**/

Reply:
Yeah, I didn't want to attempt to refine in case we didn't need it, but I outline why we need it as of now above.